### PR TITLE
Validate runtime input using zod

### PR DESCRIPTION
**Pull request by johnpmitsch**
This validates input to functions using [zod](https://www.npmjs.com/package/zod) for runtime validation of input. Without this, you can basically send anything into the function using JavaScript, or even in TypeScript you could send bad unknown input into these functions. Since we pass the arguments to the Graph API query, this makes it easy to send invalid GraphQL queries and get a not-as-obvious error back. This addition will make error handling easier (see the README section for more details) for developers and protect them from sending bad input to the Graph API. I chose zod because it's small (8kb), has 0 dependencies, and a easy-to-use syntax

I added tests to all the functions that are now validated, which helps ensure that we are setting the right zod functions on the input, in particular `.strict()` (doesn't allow unknown keys), and `.nullish()` (allows null and undefined).

Regarding typescript types, we are able to define the zod object, and then infer the typescript class from it using `z.infer` so we don't have to write the input schema twice.

Lots of the input mirrors the GQL input (I was using the codegen types as a base for the function inputs previously). As far as staying in sync with this, typescript helps us quite a bit by erroring when the `queryVariables` passed in the query don't match the codegen type we are passing in to `.query`. I also added special handling in `packages/libs/sdk/src/lib/validation/codegenDerivedValidators.ts` for the codegen string unions used in filters. We have to re-define these string unions manually with zod to use as an `.enum` as a validator. I was able to add a typescript function leveraging `never` to throw a compiler error if these don't match. Meaning, when we update the codegen generated files, we will get compiler errors if these types don't match as a reminder to update them (I have comments in this file explaining this too).

### CODE REVIEW

File: packages/libs/sdk/src/api/controllers/nfts.ts [hunk 1]
```diff
@@ -207,10 +215,6 @@ export class NftsController {
       variables: queryVariables,
     });
 
-    if (!trendingCollections?.trendingCollections?.length) {
-      return { results: [], pageInfo: emptyPageInfo };
-    }
-
```
Comment (johnpmitsch) [inline packages/libs/sdk/src/api/controllers/nfts.ts:213]: This block isn't actually doing anything now, there's no way to trigger it as trending collections should always be returned from the API and if it's not then the empty results from the API will be returned

File: packages/libs/sdk/src/api/controllers/transactions.ts [hunk 2]
```diff
@@ -50,22 +51,21 @@ import { TypedDocumentNode } from '@urql/core';
...
   constructor(
     private client: CustomUrqlClient,
     private defaultChain: ChainName = DEFAULT_CHAIN
   ) {}
 
+  @ValidateInput(balancesByWalletAddressValidator)
   async getByWallet(
     variables: TransactionsByWalletAddressInput
-  ): Promise<
-    TransactionsByWalletAddressResult | TransactionsByWalletENSResult
```
Comment (johnpmitsch) [inline packages/libs/sdk/src/api/controllers/transactions.ts:63]: These are the same result so I made it return one type (like we do in other `getByWallet` functions) for a more direct response type

File: packages/libs/sdk/src/api/graphql/queries/ethereum/mainnet/transactions/getBySearch.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { TransactionsBySearch } from '../../../fragments/getTransactionsBySearch
 
 export const EthMainnetTransactionsBySearch = gql`
   query EthMainnetTransactionsBySearch(
-    $filter: TransactionsFilterInput!,
+    $filter: TransactionsFilterInput,
```
Comment (johnpmitsch) [inline packages/libs/sdk/src/api/graphql/queries/ethereum/mainnet/transactions/getBySearch.ts:6]: I had this filter as required, but it doesn't need to be actually passed in (this is for the `transactions.getAll()` function)

**Review by AlejandroFabianCampos — APPROVED** [review]
LGTM 🔥

**Pull request merged**